File: sumofprimes.py
import PIL, math
import numpy as np

#Problem 10 sum of primes
###INEFFICIENT###
#def isPrime(number):
#	if (number <=1):
#		return False
#	for i in range(2,number):
#		if(number%i==0):
#			return False
#	return True

#def prime_factors(number):
	#primes = []
	#for i in range(number, 2, -1):
		#if(number%i==0):
			#if (isPrime(i)):
				#return i
	#return primes



def prime_factors(number):
	primes = []
	if (number <=1):
		return [1]
	a = np.ones(number)
	a[0]=0
	a[1]=0#not prime
	for i in range(2,int(math.ceil(math.sqrt(number)))):
		if(a[i]==1):
			j = i**2 #cross out all the multiples
			while(j < number):
				a[j] = False
				j+=i
	
	for k in range(2, number):
		#is prime
		if(a[k]==1):
			primes.append(k)
	return primes

out = prime_factors(2000000)
print str(sum(out))

File: sumofprimes.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

// The number to find primes up to
int n = 2000000;

// The number of primes in the primes array
int s = 0;

int *listPrimes(int num){
	int i;
	int *primes;
	int *sieve = (int *) malloc(num * sizeof(int));
	//initialize to all 1s (except 0 and 1 which are not prime)
	for(i = 2; i < num; i++)
		sieve[i] = 1;
	for(i = 2; i < ceil(sqrt(num)); i++){
		if(sieve[i] ==  1){
			int j = i * i;
			while(j < num){
				sieve[j] = 0;
				j += i;
			}
		}
	}
	
	//now check which were prime
	primes = (int *) malloc(sizeof(int));
	for(i = 2; i < num; i++){
		if(sieve[i]){
			primes[s] = i;
			s++;
			primes = (int *) realloc (primes, (s + 1) * sizeof(int));
		}
	}

	return primes;
}

int main(){
	int *p = listPrimes(n);
    long long sum = 0;
    int len = s;
    for(int i = 0; i < s; i++){
        sum += p[i];
    }
	printf("%ld\n", sum);
	return 0;
}
